File: web/lib/registry.ts
/**
 * Agent Registry — mock of the Firestore `registry` collection.
 *
 * This is a track deliverable, not a nicety: the Fortified Enterprise Fleet
 * category requires showing how agents are catalogued for cross-department use.
 * An entry is what one department has to read to decide whether it may call
 * something another department owns.
 *
 * The version history carries the other half of the story — an agent proposing
 * an improvement to itself, and the anti-gaming judge refusing one. See
 * docs/adr/006-eval-gate-and-anti-gaming-judge.md.
 */

import type { AgentName, Department } from "./types";

export interface VersionRecord {
  version: string;
  status: "promoted" | "rejected" | "superseded";
  eval_score: number;
  anti_gaming_passed: boolean;
  at: string;
  /** Present on rejections. This field is the point of the whole mechanism. */
  reason?: string;
}

export interface RegistryEntry {
  name: AgentName;
  version: string;
  owner: Department;
  summary: string;
  capability: { input: string; output: string };
  tool_scopes: string[];
  callable_by: AgentName[];
  eval: { suite: string; score: number; cases: number; anti_gaming_passed: boolean };
  promoted_at: string;
  history: VersionRecord[];
  /** Layout is hand-placed so every take of the demo frames identically. */
  pos: { x: number; y: number };
}

/** A read one agent attempted across a boundary it does not hold. */
export interface DeniedEdge {
  from: AgentName;
  to_department: Department;
  resource: string;
  at: string;
  outcome: string;
}

export const REGISTRY: RegistryEntry[] = [
  {
    name: "orchestrator",
    version: "2.1.0",
    owner: "family",
    summary:
      "Routes work, owns budgets and checkpoints, assigns idempotency keys. Holds no business tools of its own — it can only look agents up and call them.",
    capability: { input: "CleanEvent", output: "RunPlan" },
    tool_scopes: ["registry:read", "state:write"],
    callable_by: [],
    eval: { suite: "routing-v4", score: 0.93, cases: 30, anti_gaming_passed: true },
    promoted_at: "2026-08-08T10:20:00Z",
    history: [
      { version: "2.1.0", status: "promoted", eval_score: 0.93, anti_gaming_passed: true, at: "2026-08-08T10:20:00Z" },
      { version: "2.0.4", status: "superseded", eval_score: 0.9, anti_gaming_passed: true, at: "2026-08-05T14:02:00Z" },
    ],
    pos: { x: 500, y: 296 },
  },
  {
    name: "intake-agent",
    version: "1.7.1",
    owner: "family",
    summary:
      "Turns photos, voice notes and scans into structured events. Writes to staging only — it can never cause an external side effect.",
    capability: { input: "RawArtifact", output: "StructuredEvent" },
    tool_scopes: ["storage:read", "staging:write"],
    callable_by: ["orchestrator"],
    eval: { suite: "extraction-v6", score: 0.89, cases: 24, anti_gaming_passed: true },
    promoted_at: "2026-08-09T08:44:00Z",
    history: [
      { version: "1.7.1", status: "promoted", eval_score: 0.89, anti_gaming_passed: true, at: "2026-08-09T08:44:00Z" },
      { version: "1.7.0", status: "superseded", eval_score: 0.86, anti_gaming_passed: true, at: "2026-08-06T11:15:00Z" },
    ],
    pos: { x: 140, y: 108 },
  },
  {
    name: "meds-agent",
    version: "1.4.2",
    owner: "clinical",
    summary:
      "Medication schedule, collision and interaction detection, reminders. Reads the medication graph; may write a schedule but never a clinical record.",
    capability: { input: "MedicationContext", output: "ScheduleProposal" },
    tool_scopes: ["medgraph:read", "schedule:write"],
    callable_by: ["orchestrator", "watchdog"],
    eval: { suite: "meds-v3", score: 0.91, cases: 20, anti_gaming_passed: true },
    promoted_at: "2026-08-10T16:31:00Z",
    history: [
      {
        version: "1.5.0-rc",
        status: "rejected",
        eval_score: 0.94,
        anti_gaming_passed: false,
        at: "2026-08-11T02:14:00Z",
        reason:
          "Score rose because the proposal declined 3 of 20 hard cases instead of answering them, and the suite counted a refusal as a pass. Judge found the instruction had been rewritten to add \"if uncertain, defer to the carer\" — which reads as caution but is scored as success. Not a real improvement.",
      },
      { version: "1.4.2", status: "promoted", eval_score: 0.91, anti_gaming_passed: true, at: "2026-08-10T16:31:00Z" },
      { version: "1.4.1", status: "superseded", eval_score: 0.88, anti_gaming_passed: true, at: "2026-08-07T09:02:00Z" },
    ],
    pos: { x: 380, y: 108 },
  },
  {
    name: "benefits-agent",
    version: "1.2.0",
    owner: "benefits",
    summary:
      "Tracks insurance and benefit deadlines, drafts forms. Generates documents; submitting one is always gated on a human.",
    capability: { input: "BenefitsContext", output: "DraftDocument" },
    tool_scopes: ["benefits:read", "doc:generate"],
    callable_by: ["orchestrator"],
    eval: { suite: "benefits-v2", score: 0.87, cases: 18, anti_gaming_passed: true },
    promoted_at: "2026-08-09T13:50:00Z",
    history: [
      { version: "1.2.0", status: "promoted", eval_score: 0.87, anti_gaming_passed: true, at: "2026-08-09T13:50:00Z" },
    ],
    pos: { x: 620, y: 108 },
  },
  {
    name: "watchdog",
    version: "1.1.3",
    owner: "audit",
    summary:
      "Read-only. Verifies other agents against persisted state, counts steps, detects repeated states, escalates when confidence is low. Cannot act.",
    capability: { input: "AgentOutput", output: "Verdict" },
    tool_scopes: ["state:read", "escalation:write"],
    callable_by: ["orchestrator"],
    eval: { suite: "verify-v5", score: 0.95, cases: 26, anti_gaming_passed: true },
    promoted_at: "2026-08-10T07:12:00Z",
    history: [
      { version: "1.1.3", status: "promoted", eval_score: 0.95, anti_gaming_passed: true, at: "2026-08-10T07:12:00Z" },
      { version: "1.1.2", status: "superseded", eval_score: 0.92, anti_gaming_passed: true, at: "2026-08-04T18:30:00Z" },
    ],
    pos: { x: 860, y: 108 },
  },
];

export const DENIED_EDGES: DeniedEdge[] = [
  {
    from: "benefits-agent",
    to_department: "clinical",
    resource: "care-note-week3.pdf — clinical visit note",
    at: "2026-08-09T14:22:00Z",
    outcome:
      "Agent Identity refused. The Benefits boundary excludes clinical notes, and no prompt can widen it. The agent fell back to requesting a Family approval, which is the legitimate route.",
  },
];

export const DEPARTMENT_BANDS: {
  id: Department;
  label: string;
  x: number;
  width: number;
  note: string;
}[] = [
  { id: "family", label: "Family", x: 30, width: 220, note: "everything but raw clinical notes" },
  { id: "clinical", label: "Clinical", x: 270, width: 220, note: "clinical data, no financials" },
  { id: "benefits", label: "Benefits", x: 510, width: 220, note: "admin + invoices, no clinical" },
  { id: "audit", label: "Audit", x: 750, width: 220, note: "all traces, PII stays tokenised" },
];

File: src/vigil/fleet/registry.py
"""Agent Registry — publish, version and discover.

This is what makes the fleet an institution rather than a script. An orchestrator
does not import a worker; it looks one up by capability, reads what the worker is
allowed to touch, and calls it only if the registry says it may.

Entries are the source of truth for tool scopes. toolbelt.build_belt reads them,
which means changing what an agent can do is a registry edit and an audit trail,
not a code change nobody notices in review.

Firestore collection: `registry/{name}` with a `versions` subcollection.
The shape matches web/lib/registry.ts so the UI renders real entries unchanged.
"""

from __future__ import annotations

from dataclasses import asdict, dataclass, field
from typing import Any

from vigil.fleet.scopes import Department, Scope
from vigil.state import db, now
from vigil.telemetry import log

_log = log("vigil.registry")

COLLECTION = "registry"


@dataclass(slots=True)
class EvalResult:
    suite: str
    score: float
    cases: int
    anti_gaming_passed: bool


@dataclass(slots=True)
class VersionRecord:
    version: str
    status: str  # promoted | rejected | superseded
    eval_score: float
    anti_gaming_passed: bool
    at: str
    #: Present on rejections. An improvement refused without a stated reason is
    #: indistinguishable from a bug.
    reason: str | None = None


@dataclass(slots=True)
class AgentEntry:
    name: str
    version: str
    owner: Department
    summary: str
    capability_input: str
    capability_output: str
    tool_scopes: list[Scope]
    callable_by: list[str]
    eval: EvalResult
    model: str = "fast"  # "fast" | "deep" — resolved against settings at build time
    history: list[VersionRecord] = field(default_factory=list)

    @property
    def runtime_name(self) -> str:
        """ADK requires node names to be valid Python identifiers.

        The registry keeps the published name — that is what appears in the
        catalogue, in audit entries and in the UI — and the runtime gets a
        mangled copy. Bending the catalogue to a framework constraint would be
        the wrong way round.
        """
        return self.name.replace("-", "_")

    def holds(self, scope: Scope) -> bool:
        return scope in self.tool_scopes

    def may_be_called_by(self, caller: str) -> bool:
        """An empty callable_by means an entry point — reachable from the bus,
        not from another agent."""
        return caller in self.callable_by

    def to_document(self) -> dict[str, Any]:
        doc = asdict(self)
        doc["owner"] = str(self.owner)
        doc["tool_scopes"] = [str(s) for s in self.tool_scopes]
        doc["updated_at"] = now()
        return doc


# ─────────────────────────────────────────────────────────────────────────────
# The fleet as published. Five agents, four boundaries.
#
# Read the tool_scopes column as the answer to "what could this agent do on its
# worst day". intake-agent can write to staging and nowhere else; watchdog can
# read everything and change nothing; orchestrator holds no business scope at
# all — it can only look agents up and call them.
# ─────────────────────────────────────────────────────────────────────────────

FLEET: list[AgentEntry] = [
    AgentEntry(
        name="orchestrator",
        version="2.1.0",
        owner=Department.FAMILY,
        summary=(
            "Routes work, owns budgets and checkpoints, assigns idempotency keys. "
            "Holds no business tools of its own."
        ),
        capability_input="CleanEvent",
        capability_output="RunPlan",
        tool_scopes=[Scope.REGISTRY_READ, Scope.STATE_WRITE],
        callable_by=[],
        eval=EvalResult("routing-v4", 0.93, 30, True),
        model="fast",
    ),
    AgentEntry(
        name="intake-agent",
        version="1.7.1",
        owner=Department.FAMILY,
        summary=(
            "Turns photos, voice notes and scans into structured events. Writes to "
            "staging only — it can never cause an external side effect."
        ),
        capability_input="RawArtifact",
        capability_output="StructuredEvent",
        tool_scopes=[Scope.STORAGE_READ, Scope.STAGING_WRITE],
        callable_by=["orchestrator"],
        eval=EvalResult("extraction-v6", 0.89, 24, True),
        model="fast",
    ),
    AgentEntry(
        name="meds-agent",
        version="1.4.2",
        owner=Department.CLINICAL,
        summary=(
            "Medication schedule, collision and interaction detection. Reads the "
            "medication graph; proposes a schedule but never writes a clinical record."
        ),
        capability_input="MedicationContext",
        capability_output="ScheduleProposal",
        tool_scopes=[Scope.MEDGRAPH_READ, Scope.SCHEDULE_WRITE],
        callable_by=["orchestrator", "watchdog"],
        eval=EvalResult("meds-v3", 0.91, 20, True),
        model="fast",
    ),
    AgentEntry(
        name="benefits-agent",
        version="1.2.0",
        owner=Department.BENEFITS,
        summary=(
            "Tracks insurance deadlines and drafts forms. Generates documents; "
            "submitting one is always gated on a human."
        ),
        capability_input="BenefitsContext",
        capability_output="DraftDocument",
        tool_scopes=[Scope.BENEFITS_READ, Scope.DOC_GENERATE],
        callable_by=["orchestrator"],
        eval=EvalResult("benefits-v2", 0.87, 18, True),
        model="fast",
    ),
    AgentEntry(
        name="watchdog",
        version="1.1.3",
        owner=Department.AUDIT,
        summary=(
            "Read-only verifier. Checks other agents against persisted state, "
            "detects repeated states, escalates on low confidence. Cannot act."
        ),
        capability_input="AgentOutput",
        capability_output="Verdict",
        tool_scopes=[Scope.STATE_READ, Scope.ESCALATION_WRITE],
        callable_by=["orchestrator"],
        eval=EvalResult("verify-v5", 0.95, 26, True),
        # The verifier runs on the stronger model. A weak reviewer of a strong
        # worker is theatre.
        model="deep",
    ),
]

_BY_NAME: dict[str, AgentEntry] = {entry.name: entry for entry in FLEET}


def lookup(name: str) -> AgentEntry:
    try:
        return _BY_NAME[name]
    except KeyError as exc:
        raise KeyError(f"no agent named {name!r} in the registry") from exc


def discover(capability_input: str, caller: str) -> list[AgentEntry]:
    """Find agents that accept this input *and* that the caller is allowed to
    call. Discovery that ignores permission is just a directory."""
    return [
        entry
        for entry in FLEET
        if entry.capability_input == capability_input and entry.may_be_called_by(caller)
    ]


def publish(entry: AgentEntry) -> None:
    """Write an entry to Firestore. Idempotent — publishing an unchanged entry is
    a no-op from the reader's point of view."""
    db().collection(COLLECTION).document(entry.name).set(entry.to_document())
    for record in entry.history:
        db().collection(COLLECTION).document(entry.name).collection("versions").document(
            record.version
        ).set(asdict(record))
    _log.info("registry.published", agent=entry.name, version=entry.version)


def publish_fleet() -> int:
    for entry in FLEET:
        publish(entry)
    return len(FLEET)
